compiler/rustc_codegen_llvm/src/llvm_util.rs

```diff
@@ -389,36 +383,90 @@ pub fn llvm_global_features(sess: &Session) -> Vec<String> {
         Some(_) | None => {}
     };
 
-    let filter = |s: &str| {
-        if s.is_empty() {
-            return vec![];
-        }
-        let feature = if s.starts_with('+') || s.starts_with('-') {
-            &s[1..]
-        } else {
-            return vec![s.to_string()];
-        };
-        // Rustc-specific feature requests like `+crt-static` or `-crt-static`
-        // are not passed down to LLVM.
-        if RUSTC_SPECIFIC_FEATURES.contains(&feature) {
-            return vec![];
-        }
-        // ... otherwise though we run through `to_llvm_feature` feature when
-        // passing requests down to LLVM. This means that all in-language
-        // features also work on the command line instead of having two
-        // different names when the LLVM name and the Rust name differ.
-        to_llvm_feature(sess, feature).iter().map(|f| format!("{}{}", &s[..1], f)).collect()
-    };
-
     // Features implied by an implicit or explicit `--target`.
-    features.extend(sess.target.features.split(',').flat_map(&filter));
+    features.extend(
+        sess.target
+            .features
+            .split(',')
+            .filter(|v| !v.is_empty() && backend_feature_name(v).is_some())
+            .map(String::from),
+    );
 
     // -Ctarget-features
-    features.extend(sess.opts.cg.target_feature.split(',').flat_map(&filter));
+    let supported_features = supported_target_features(sess);
+    features.extend(
+        sess.opts
+            .cg
+            .target_feature
+            .split(',')
+            .filter_map(|s| {
+                let enable_disable = match s.chars().next() {
+                    None => return None,
+                    Some(c @ '+' | c @ '-') => c,
+                    Some(_) => {
+                        let mut diag = sess.struct_warn(&format!(
+                            "unknown feature specified for `-Ctarget-feature`: `{}`",
+                            s
+                        ));
+                        diag.note("features must begin with a `+` to enable or `-` to disable it");
+                        diag.emit();
+                        return None;
+                    }
+                };
+
+                let feature = backend_feature_name(s)?;
+                // Warn against use of LLVM specific feature names on the CLI.
+                if !supported_features.iter().any(|&(v, _)| v == feature) {
+                    let rust_feature = supported_features.iter().find_map(|&(rust_feature, _)| {
+                        let llvm_features = to_llvm_features(sess, rust_feature);
```

Comment:
Should we be doing fuzzy searching in `to_llvm_features` for the suggestion given bellow?

Reply:
Possibly. My primary goal was to direct users from using LLVM names towards using Rust names, and much less so to handle typos better. I do agree however that a fuzzy suggestions here would be ideal. Even better if we make them work for more than just `-Ctarget-features`.